ref(billing): Convert InvoiceItemType to dynamic type union

dashed opened the pull request:
Closes https://linear.app/getsentry/issue/BIL-969/dynamic-invoiceitemtypefe

Replaces the hardcoded `InvoiceItemType` enum with a dynamically-generated type union that automatically syncs with `DATA_CATEGORY_INFO`. This eliminates the need for manual enum updates whenever new billing categories are added.

The implementation follows the existing `EventType` pattern and uses TypeScript mapped types to generate `ondemand_*` and `reserved_*` invoice item types from all billed categories. Added CamelToSnake helper type to convert DataCategory's camelCase values to backend's snake_case format.

This change also fixes 4 missing invoice types (ondemand_profile_duration_ui, ondemand_log_bytes, ondemand_prevent_users, reserved_profile_duration_ui) and prevents future category additions from causing missing type bugs.

dashed committed:
ref(billing): Convert InvoiceItemType to dynamic type union

Replaces the hardcoded InvoiceItemType enum with a dynamically-generated
type union that automatically syncs with DATA_CATEGORY_INFO. This eliminates
the need for manual enum updates whenever new billing categories are added.

The implementation follows the existing EventType pattern and uses TypeScript
mapped types to generate ondemand_* and reserved_* invoice item types from
all billed categories. Added CamelToSnake helper type to convert DataCategory's
camelCase values to backend's snake_case format.

This change fixes 4 missing invoice types (ondemand_profile_duration_ui,
ondemand_log_bytes, ondemand_prevent_users, reserved_profile_duration_ui)
and prevents future category additions from causing missing type bugs.

Refs BIL-969

isabellaenriquez commented on static/gsApp/types/index.tsx:
```diff
@@ -649,51 +650,75 @@ export type InvoiceItem = BaseInvoiceItem & {
+ * Static invoice item types that are not tied to data categories.
+ * These must be manually maintained but change infrequently.
+ */
+type StaticInvoiceItemType =
```
we'll just need to add `activated_seer_users`

isabellaenriquez reviewed (commented):
very nice

just one thing to address, everything else is nits

dashed committed:
fix(billing): Add missing reserved_seer_users and activated_seer_users types

Adds two special-case invoice item types that were missing from the
initial dynamic type implementation:

- reserved_seer_users: Used for PREVENT_USER category reserved billing.
  Backend maps PREVENT_USER to this instead of reserved_prevent_users
  due to naming conventions that will be unified later.

- activated_seer_users: Used for activation-based billing model for
  Prevent users. This is a newer billing model that works similarly
  to PAYG but is tracked separately.

These types don't follow the standard category naming pattern due to
backend special-case logic in invoice_item_type.py for_reserved() and
to_activated_attr() methods.

Refs BIL-969

dashed committed:
refactor(billing): Organize invoice item types into semantic sub-categories

- Created const arrays for credit, fee, seer, and legacy invoice types
- Updated getCredits and getFees to use const arrays instead of hardcoded strings
- Composed StaticInvoiceItemType from organized sub-types
- Enables runtime usage in filters while maintaining type safety

This eliminates duplication between type definitions and runtime values,
providing a single source of truth for invoice item type categorization.

dashed committed:
fix(billing): Fix CamelToSnake to handle consecutive capitals correctly

The previous implementation incorrectly handled consecutive capital letters
by inserting underscores between each letter. For example, 'profileDurationUI'
was converted to 'profile_duration_u_i' instead of 'profile_duration_ui'.

The fix tracks the previous character's case state and looks ahead at the next
character to determine if we're at a word boundary:
- Consecutive capitals (like 'UI') are kept together without underscores
- Transitions from lowercase to uppercase still insert underscores
- Transitions from uppercase to lowercase insert underscores before the capital

This ensures backend type alignment:
- ondemand_profile_duration_ui (was: ondemand_profile_duration_u_i)
- reserved_profile_duration_ui (was: reserved_profile_duration_u_i)

dashed commented on static/gsApp/utils/billing.tsx:
```diff
@@ -804,12 +805,12 @@ export function getCredits({
+        'subscription_credit',
+        'credit_applied', // TODO(isabella): This is deprecated and replaced by BALANCE_CHANGE
+        'discount',
+        'recurring_discount',
```
it's separated out now.

dashed committed:
fix(eslint): Remove unnecessary type assertions

Removed unnecessary type assertions that were flagged by ESLint:
- inputField.tsx: Removed `as any` from e.target.value
- numberField.tsx: Removed `as any` from e.target.value
- onboarding.tsx: Removed `as number` assertions from stepper onClick handler

These assertions were unnecessary because TypeScript already infers the correct
types from the event handlers and component props.

dashed committed:
fix(knip): Remove unused exports from invoice item types

Removed export keywords from unused constants and types:
- SEER_INVOICE_ITEM_TYPES → _SEER_INVOICE_ITEM_TYPES (only used for type derivation)
- LEGACY_INVOICE_ITEM_TYPES → _LEGACY_INVOICE_ITEM_TYPES (only used for type derivation)
- CreditInvoiceItemType, FeeInvoiceItemType, etc. (only used internally to compose StaticInvoiceItemType)

Only CREDIT_INVOICE_ITEM_TYPES and FEE_INVOICE_ITEM_TYPES remain exported as
they are actively used in billing.tsx for runtime filtering.

dashed committed:
fix(typescript): Fix TypeScript errors caused by DOMAttributes change

Root cause: Changing DOMAttributes<T> to DOMAttributes<_T> broke React's
module augmentation, causing all React components to lose standard props.

Fixes:
1. Reverted DOMAttributes<_T> back to DOMAttributes<T> with eslint-disable
   - Module augmentation requires matching generic parameter names
   - Added eslint-disable-next-line for the unused-vars false positive

2. Fixed inputField and numberField type assertions
   - Restored (e.target as HTMLInputElement) for onKeyDown handlers
   - e.target is EventTarget, needs cast to access .value property

3. Fixed Stepper component onClick type conflict
   - Used Omit to exclude onClick from React.HTMLAttributes
   - Prevents intersection type conflict between DOM and custom onClick

4. Removed unused DataCategory import from utils.spec.tsx

All TypeScript checks now pass with exit code 0.

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

isabellaenriquez reviewed (approved):
lgtm, love the random type fixes lol

dashed commented on static/gsApp/types/index.tsx:
```diff
@@ -649,51 +650,75 @@ export type InvoiceItem = BaseInvoiceItem & {
+  | 'balance_change'
+  | 'cancellation_fee'
+  | 'attachments' // Legacy: AM1 plans
+  | 'transactions' // Legacy: AM1 plans
```
Just their old invoices.
